File: rest-api/src/maven/maven.go
package maven

import (
	"fmt"
	"log"
	"net/http"
	"net/url"
	"strings"
)

const repoBaseURL = "https://repo1.maven.org/maven2"

var cache = make(map[string]bool)

type Artifact struct {
	GroupId    string `json:"groupId"`
	ArtifactId string `json:"artifactId"`
	Version    string `json:"version"`
}

func NewArtifact(coordinates string) (*Artifact, error) {

	parts := strings.Split(coordinates, ":")
	if len(parts) != 3 {
		err := fmt.Errorf("invalid artifact coordinates: '%s'", coordinates)
		return nil, err
	}

	artifact := Artifact{
		GroupId:    parts[0],
		ArtifactId: parts[1],
		Version:    parts[2],
	}

	return &artifact, nil
}

// Exists Check if the artifact exists in Maven Central.
func (artifact *Artifact) Exists() (bool, error) {

	// check in-memory cache
	cacheKey := artifact.ToCoordinates()
	exists, cached := cache[cacheKey]
	if cached {
		if exists {
			log.Println("artifact found:", cacheKey, "(cached)")
			return true, nil
		} else {
			log.Println("artifact not found:", cacheKey, "(cached)")
			return false, nil
		}
	}

	// get artifact URL for given coordinates
	artifactURL := artifact.GetURL()
	log.Println("artifact URL:", artifactURL)

	// check if artifact URL is valid
	resp, err := http.Head(artifactURL)
	if err != nil {
		return false, err
	}

	// return result based on status code
	statusCode := resp.StatusCode
	if statusCode == http.StatusOK {
		log.Println("artifact found:", cacheKey)
		cache[cacheKey] = true
		return true, nil
	} else if statusCode == http.StatusNotFound {
		log.Println("artifact not found:", cacheKey)
		cache[cacheKey] = false
		return false, nil
	} else {
		err := fmt.Errorf("unexpected status code: %d", statusCode)
		return false, err
	}
}

// GetURL Generate artifact URL based on coordinates.
func (artifact *Artifact) GetURL() string {

	var artifactURL strings.Builder
	artifactURL.WriteString(repoBaseURL)
	for _, path := range strings.Split(artifact.GroupId, ".") {
		artifactURL.WriteString("/")
		artifactURL.WriteString(url.PathEscape(path))
	}
	artifactURL.WriteString("/")
	artifactURL.WriteString(url.PathEscape(artifact.ArtifactId))
	artifactURL.WriteString("/")
	artifactURL.WriteString(url.PathEscape(artifact.Version))
	artifactURL.WriteString("/")
	artifactURL.WriteString(url.PathEscape(artifact.ArtifactId))
	artifactURL.WriteString("-")
	artifactURL.WriteString(url.PathEscape(artifact.Version))
	artifactURL.WriteString(".jar")

	return artifactURL.String()
}

func (artifact *Artifact) ToCoordinates() string {
	return fmt.Sprintf("%s:%s:%s", artifact.GroupId, artifact.ArtifactId, artifact.Version)
}

File: sam-api/shared/src/main/java/org/jarhc/online/clients/Maven.java
package org.jarhc.online.clients;

import com.amazonaws.xray.AWSXRay;
import com.amazonaws.xray.entities.Subsegment;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jarhc.online.Artifact;

public class Maven {

	private static final Logger logger = LogManager.getLogger();

	private static final String BASE_URL = "https://repo1.maven.org/maven2";
	private static final int STATUS_OK = 200;
	private static final int STATUS_NOT_FOUND = 404;

	private static final Map<String, Boolean> CACHE = new HashMap<>();

	private final int timeout;

	/**
	 * Create a Maven client.
	 *
	 * @param timeout Timeout in milliseconds.
	 */
	public Maven(int timeout) {
		this.timeout = timeout;
	}

	public boolean exists(Artifact artifact) throws MavenException {
		try (Subsegment xray = AWSXRay.beginSubsegment("Maven.exists")) {
			var coordinates = artifact.toCoordinates();
			xray.putAnnotation("artifact", coordinates);

			// check if artifact result exists in cache
			if (CACHE.containsKey(coordinates)) {
				boolean found = CACHE.get(coordinates);
				xray.putAnnotation("cached", found);
				if (found) {
					logger.debug("Artifact found : {} (cached)", coordinates);
					return true;
				} else {
					logger.debug("Artifact not found : {} (cached)", coordinates);
					return false;
				}
			}

			// send HTTP HEAD request to Maven Central repository URL
			int statusCode = sendRequest("HEAD", artifact, null, xray);

			// interpret response status code
			switch (statusCode) {
				case STATUS_OK:
					logger.debug("Artifact found : {}", coordinates);
					CACHE.put(coordinates, true);
					return true;
				case STATUS_NOT_FOUND:
					logger.debug("Artifact not found : {}", coordinates);
					CACHE.put(coordinates, false);
					return false;
				default:
					throw new MavenException("Unexpected status code: " + statusCode);
			}
		}
	}

	public boolean download(Artifact artifact, File file) throws MavenException {
		try (Subsegment xray = AWSXRay.beginSubsegment("Maven.download")) {
			var coordinates = artifact.toCoordinates();
			xray.putAnnotation("artifact", coordinates);

			// send HTTP GET request to Maven Central repository URL
			Consumer<HttpURLConnection> processor = connection -> {
				try {
					saveResponseToFile(connection, file);
				} catch (IOException e) {
					throw new UncheckedIOException(e);
				}
			};
			int statusCode = sendRequest("GET", artifact, processor, xray);

			// interpret response status code
			switch (statusCode) {
				case STATUS_OK:
					logger.debug("Artifact downloaded : {} -> {}", coordinates, file.getAbsolutePath());
					return true;
				case STATUS_NOT_FOUND:
					logger.warn("Artifact not found : {}", coordinates);
					return false;
				default:
					throw new MavenException("Unexpected status code: " + statusCode);
			}
		}
	}

	private int sendRequest(String method, Artifact artifact, Consumer<HttpURLConnection> processor, Subsegment xray) throws MavenException {

		HttpURLConnection connection = null;
		try {

			// prepare URL
			var url = new URL(BASE_URL + artifact.toURLPath());
			logger.debug("Artifact URL: {}", url);

			// prepare HTTP connection
			connection = (HttpURLConnection) url.openConnection();
			connection.setRequestMethod(method);
			connection.setConnectTimeout(timeout);
			connection.setReadTimeout(timeout);
			connection.setDoOutput(false);
			connection.setDoInput(true);

			// send request and get response
			connection.connect();

			// get response status code
			int statusCode = connection.getResponseCode();
			logger.debug("Status code: {}", statusCode);
			xray.putAnnotation("statusCode", statusCode);

			if (statusCode == STATUS_OK && processor != null) {
				processor.accept(connection);
			}

			return statusCode;

		} catch (Exception e) {
			xray.addException(e);
			throw new MavenException("HTTP error", e);
		} finally {
			// close connection
			if (connection != null) {
				connection.disconnect();
			}
		}
	}

	private static void saveResponseToFile(HttpURLConnection connection, File file) throws IOException {

		// make sure that parent directory exists
		File directory = file.getParentFile();
		if (!directory.isDirectory()) {
			boolean created = directory.mkdirs();
			if (!created) {
				throw new IOException("Error creating directory: " + directory.getAbsolutePath());
			}
		}

		// save response body to file
		try (InputStream inputStream = connection.getInputStream()) {
			try (FileOutputStream outputStream = new FileOutputStream(file)) {
				inputStream.transferTo(outputStream);
			}
		}
	}

}
